### PR TITLE
Fix/search

### PR DESCRIPTION
**Pull request by tbrezot**
- prevent loop when searching through a cyclic keyword graph.
- implement `Copy` for `Uid`
- simplify `EntryTableValue::unchain()`
- delete `EntryTable::unchain()` which may not be a good choice, please give me your opinion on it :)

### CODE REVIEW

File: src/in_memory_example.rs [hunk 1]
```diff
@@ -70,35 +69,24 @@ impl<const UID_LENGTH: usize> FindexInMemory<UID_LENGTH> {
...
-            let results = results
-                .get(keyword)
-                .ok_or_else(|| {
-                    ExampleError(format!(
-                        "Cannot find keyword {keyword:?} in search results {results:?}"
-                    ))
-                })
-                .unwrap();
-            assert!(results.contains(&IndexedValue::NextKeyword(Keyword::from("robert"))));
-        }
```
Comment (ThibaudDauce) [inline src/in_memory_example.rs:97]: Did you add back this test somewhere else?

File: tests/test_in_memory.rs [hunk 2]
```diff
@@ -288,12 +277,11 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
         // search doe
         let doe_search = findex
             .search(
-                &HashSet::from_iter(vec![doe_keyword.clone()]),
                 &master_key,
                 &new_label,
-                usize::MAX,
-                0,
-                0,
+                &HashSet::from_iter(vec![doe_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:282]:
```suggestion
                &HashSet::from_iter([doe_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 3]
```diff
@@ -219,57 +214,51 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
...
-    findex.set_check_progress_callback_next_keyword(true);
     let rob_search = findex
         .search(
-            &HashSet::from_iter(vec![rob_keyword.clone()]),
             &master_key,
             &label,
-            usize::MAX,
-            0,
-            0,
+            &HashSet::from_iter(vec![rob_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:257]:
```suggestion
            &HashSet::from_iter([rob_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 4]
```diff
@@ -219,57 +214,51 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
...
     // search robert (no change)
     let robert_search = findex
         .search(
-            &HashSet::from_iter(vec![robert_keyword.clone()]),
             &master_key,
             &label,
-            usize::MAX,
-            0,
-            0,
+            &HashSet::from_iter(vec![robert_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:231]:
```suggestion
            &HashSet::from_iter([robert_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 5]
```diff
@@ -219,57 +214,51 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
     let jane_keyword = Keyword::from("jane");
     let jane_search = findex
         .search(
-            &HashSet::from_iter(vec![jane_keyword.clone()]),
             &master_key,
             &label,
-            usize::MAX,
-            0,
-            0,
+            &HashSet::from_iter(vec![jane_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:219]:
```suggestion
            &HashSet::from_iter([jane_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 6]
```diff
@@ -130,67 +132,60 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
...
     // search rob with graph search
     let rob_search = findex
         .search(
-            &HashSet::from_iter(vec![rob_keyword.clone()]),
             &master_key,
             &label,
-            usize::MAX,
-            usize::MAX,
-            0,
+            &HashSet::from_iter(vec![rob_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:186]:
```suggestion
            &HashSet::from_iter([rob_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 7]
```diff
@@ -130,67 +132,60 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
...
-    findex.set_check_progress_callback_next_keyword(true);
     let rob_search = findex
         .search(
-            &HashSet::from_iter(vec![rob_keyword.clone()]),
             &master_key,
             &label,
-            usize::MAX,
-            0,
-            0,
+            &HashSet::from_iter(vec![rob_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:174]:
```suggestion
            &HashSet::from_iter([rob_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 8]
```diff
@@ -130,67 +132,60 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
...
     // cannot find robert with wrong label
     let robert_search = findex
         .search(
-            &HashSet::from_iter(vec![robert_keyword.clone()]),
             &master_key,
             &Label::random(&mut rng),
-            usize::MAX,
-            0,
-            0,
+            &HashSet::from_iter(vec![robert_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:149]:
```suggestion
            &HashSet::from_iter([robert_keyword.clone()]),
```

File: tests/test_in_memory.rs [hunk 9]
```diff
@@ -130,67 +132,60 @@ async fn test_findex() -> Result<(), Error<ExampleError>> {
     // search robert
     let robert_search = findex
         .search(
-            &HashSet::from_iter(vec![robert_keyword.clone()]),
             &master_key,
             &label,
+            &HashSet::from_iter(vec![robert_keyword.clone()]),
```
Comment (ThibaudDauce) [inline tests/test_in_memory.rs:137]:
```suggestion
            &HashSet::from_iter([robert_keyword.clone()]),
```

File: src/entry_table.rs [hunk 10]
```diff
@@ -226,42 +226,26 @@ impl<const UID_LENGTH: usize, const KWI_LENGTH: usize> EntryTableValue<UID_LENGT
...
-        for _ in 0..max_results {
-            // Add the new Chain Table UID to the map.
-            entry.push(current_chain_table_uid.clone());
-
-            // Return if we found the UID stored in the Entry Table value.
-            if Some(&current_chain_table_uid) == self.chain_table_uid.as_ref() {
-                break;
-            }
+        kwi_uid: &KmacKey,
+        max_uid_per_chain: usize,
```
[the remark this answers is not in the document]
Comment (tbrezot) [reply src/entry_table.rs:227]: good catch

File: src/in_memory_example.rs (hunk 1 above)
Comment (tbrezot) [reply src/in_memory_example.rs:97]: No, I didn't find it that useful...

Comment (ThibaudDauce) [reply src/in_memory_example.rs:97]: It checks that the progress callback is working no? Without it we can remove the call to the progress callback?

**Review by ThibaudDauce — APPROVED** [review]

File: src/in_memory_example.rs (hunk 1 above)
Comment (tbrezot) [reply src/in_memory_example.rs]: I have added it back.

**Pull request merged**